gtp_engine.py

```diff
@@ -0,0 +1,197 @@
+    if not msg:
+        return None, None, None
+    parts = [x for x in msg.split(" ") if x]
+    if len(parts) > 1 and parts[0].isdigit():
+        msg_id = parts[0]
+        parts = parts[1:]
+    else:
+        msg_id = None
+    return msg_id, parts[0], parts[1:]
+
+
+def _print_msg(result, msg_id, msg):
+    if msg_id:
+        msg_id = " {}".format(msg_id)
+    else:
+        msg_id = ""
+    if isinstance(msg, bool):
+        msg = "true" if msg else "false"
+    if msg:
+        msg = " {}".format(msg)
+    else:
+        msg = ""
+    print("{}{}{}\n".format(result, msg_id, msg), flush=True)
+
+
+def _print_error(msg_id, msg):
+    _print_msg("?", msg_id, msg)
+
+
+def _print_success(msg_id, msg):
+    _print_msg("=", msg_id, msg)
+
+
+def _convert_args(handler, args):
+    """Convert a list of command arguments to types specified by the handler.
+
+    Args:
+      handler: a command handler function.
+      args: the list of string arguments to pass to handler.
+
+    Returns:
+      A new list containing `args` that have been converted to the expected type
+      for `handler`. For each function parameter of `handler` that has either an
+      explicit type annotation or a non-None default value, the corresponding
+      element in `args` is converted to that type.
+    """
+
+    args = list(args)
+    params = inspect.signature(handler).parameters
+    for i, (arg, name) in enumerate(zip(args, params)):
+        default = params[name].default
+        annotation = params[name].annotation
+
+        if annotation != inspect.Parameter.empty:
+            if isinstance(annotation, type) and annotation != str:
+                # The parameter is annotated with a type that isn't str: convert
+                # the arg to that type.
+                args[i] = annotation(arg)
+        elif default != inspect.Parameter.empty:
+            if default is not None and not isinstance(default, str):
+                # The parameter has a default value that isn't None or a str:
+                # convert the arg to the default value's type.
+                args[i] = type(default)(arg)
+
+    return args
+
+
+class Engine(object):
+    """A simple GTP engine.
+
+    The engine by itself doesn't do anything: clients must register command
+    handler objects using `add_cmd_handler`.
+    """
+
+    def __init__(self):
+        self.cmds = {}
+
+    def add_cmd_handler(self, handler_obj):
+        """Registers a new command handler object.
+
+        All methods on `handler_obj` whose name starts with "cmd_" are
+        registered as a GTP command. For example, the method cmd_genmove will
+        be invoked when the engine receives a genmove command.
+
+        Args:
+          handler_obj: the handler object to register.
+        """
+        for field in dir(handler_obj):
+            if field.startswith("cmd_"):
+                cmd = field[4:]
+                fn = getattr(handler_obj, field)
+                self.cmds[cmd] = fn
```

Comment:
maybe print a warning if a cmd is defined twice? The override behavior seems to be done by later cmd handlers overwrite earlier ones.

Reply:
done.
good suggestion, thanks